docs/internal/types.md

```diff
@@ -0,0 +1,246 @@
+ * Capability rights: A Verona pointer is a *capability*: an unforgeable token of authority that permits access to an object. 
+   The set of rights that a capability holds are associated with a type, for example. `iso`, `mut`, `imm` and `readonly`.
+ * Unions: The concrete type of the object must fulfil all of the constraints of at least one out of two or more type expressions, for example the union type `(A | B)` must fulfil all of the requirements of either `A` or `B`.
+   Unions may refer to concrete types (for example, if `A` and `B` are classes then a variable of type `A | B` may be a reference to an instance of either class) or other type expressions, in which case the variable may refer to any concrete type that meets the requirements of either set of constraints.
+ * Intersections: The concrete type of the object must fulfil all of the constraints of two or more type expressions, for example the union type `(A & B)` must fulfil all of the requirements of both `A` and `B`, which may be interfaces or other type expressions.
+	   No type satisfies the union type of two concrete types, so if `A` and `B` are both class types then `A` & `B` is a type that is not possible to satisfy.
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an the sentinel object in a region.
+   Every region has a single sentinel that dominates all of the live objects in the region.
+   Exactly one `iso` pointer exists for any region and the behaviour accessing any mutable object must hold the `iso` pointer for the sentinel object of the corresponding region.
+ * `mut`, or `mutable`, means this object can be modified.
+   All mutable objects are contained by a region, the behaviour modifying the object must own the `iso` pointer that refers to the sentinel object of the region.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time. You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+  * `readonly` defines local immutability.
+    No object reached via a `readonly` capability may be modified via that the object is not guaranteed to be immutable.
+    `mut` and `readonly` capabilities to the same object may be visible at the same time.
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+A `freeze` operation consumes an `iso`, destroying the only copy of the capability that is required to mutate any object in a region, and returns an `imm` capability.
+There is no corresponding `unfreze` operation to consume an `imm` capability and provide any kind of mutable capability.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
```

Comment:
Possibly 'may have'?  

I'm possibly confused by 'run-time' vs 'runtime' here:  'run-time' is an adjective phrase and modifies semantics, so 'run-time semantics' refers to semantics that apply at the time the program runs.  'Runtime' is a noun that refers to the library that provides services to the program at run time, so 'runtime semantics', to me, means operations that the runtime provides.

Casting from `iso` to `mut` or `mut` to `readonly`, for example, change the set of operations that are permitted with the resulting capability (and are monotonic - you can't cast back) but (probably) don't involve the runtime doing anything.